### DISCUSSION TITLE
Load Pyodide correctly in React with "strict" enabled?

### DISCUSSION THREAD

**Discussion by vanweric** [pyodide/pyodide#4957] [Q&A]
While developing locally (npm start), react defaults to "strict mode" which stress tests all modules by loading them, unloading them, and then loading them again.

My code is architected such that the end of loading pyodide starts loading app specific python state and functions.  I'm having difficulty understanding how I should trigger and gate this loading so that I guarantee a coherent state once everything is up.

Disabling strict mode works as a temporary solution, but I'd appreciate help on how to do this correctly.

### Loading Pyodide:
```
import { useEffect, useState } from 'react';

const usePyodide = () => {
  const [pyodide, setPyodide] = useState(null);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(null);

  useEffect(() => {
    const loadPyodide = async () => {
      try {
        console.log("Start Pyodide Load");
        const script = document.createElement('script');
        //script.src = 'https://cdn.jsdelivr.net/pyodide/v0.23.2/full/pyodide.js';
        script.src = 'https://cdn.jsdelivr.net/pyodide/v0.26.1/full/pyodide.js';
        script.onload = async () => {
          const pyodide = await window.loadPyodide({
            //indexURL: 'https://cdn.jsdelivr.net/pyodide/v0.23.2/full/',
            indexURL: 'https://cdn.jsdelivr.net/pyodide/v0.26.1/full/',
          });
          setPyodide(pyodide);
          setLoading(false);
          console.log("End Pyodide Load");
        };
        document.head.appendChild(script);
      } catch (err) {
        console.log(err);
        setError(err);
        setLoading(false);
      }
    };

    loadPyodide();
  }, []);

  return { pyodide, loading, error };
};

export default usePyodide;
```
### Trigger Setup:
```
  useEffect(() => {
    const setupPythonFunctions = async () => {
      setCompiling(true);
      if (!loading) {
        console.log("Start Loadup");
        console.log(pyodide.runPython(await (await fetch("./mycode.py")).text()));
        console.log(pyodide.runPython("function_from_my_code()"));
        console.log("End Loadup");
        setCompiling(false);
      }
    }
    setupPythonFunctions();
  }, [loading]);
```
### useage

```
  const runPythonCode = async () => {
    if (pyodide) {
      setCompiling(true);
      const locals = pyodide.toPy({ arg_text: variable_from_elsewhere })
      const result = pyodide.runPython('function_loaded(arg_text)', { locals: locals });
      setOutput(result);
      setCompiling(false);

    }
  };
        <button className="btn btn-primary" type="button" disabled={compiling} onClick={runPythonCode}>
```

## Other Tests
I've reproduced the behavior I'm seeing by using a simple test function and eliminating all of my python code
`def square(x): return x*x`
I believe this eliminates my python code as the culprit.

I've triggered the function loading from a button instead of a use effect, and I can trigger the behavior by hitting the load function button after the first pyodide load starts but before the second pyodide load.  If I wait it works as expected.

I think this strongly implies that it is losing state on pyodide reload (reasonable behavior).  I just don't know the correct way to trigger loading up my python context safely in the face of that optional reload.

Thanks in advance!

**Comment by ryanking13** [comment]
> losing state on pyodide reload (reasonable behavior)

Yes, each pyodide instance is independent.

> I just don't know the correct way to trigger loading up my python context safely in the face of that optional reload.

Well, I think this is rather a react issue, not Pyodide issue, but here is my best guess (based on by basic react knowledge)... useEffect should be called again when the "second pyodide load" happens, for now `pyodide` object is not in the useEffect dependencies so maybe it is not re-triggered?
